File: product/transform/contracts.py
"""Transform request/result contracts — provider-agnostic, mobile-facing shape."""
from __future__ import annotations

from dataclasses import asdict, dataclass, field
from typing import Any


STYLES = ("gummy", "clay", "plush", "glossy")


@dataclass
class TransformOptions:
    """Client-safe options. Prompt / provider knobs stay server-side."""

    size: int = 1024
    dry_run: bool = False
    # Optional recognition override (tests / future VLM). Never required from mobile.
    recognition: dict[str, Any] | None = None

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


@dataclass
class TransformRequest:
    """Inbound transform contract for POST /v1/transform.

    No provider names, prompts, style sheets, or API keys.
    """

    style: str
    doodle_png: bytes | None = None
    doodle_path: str | None = None
    strokes: dict[str, Any] | None = None
    client_doodle_id: str | None = None
    options: TransformOptions = field(default_factory=TransformOptions)

    def __post_init__(self) -> None:
        style = (self.style or "").lower().strip()
        if style not in STYLES:
            raise ValueError(f"style must be one of {STYLES}; got {self.style!r}")
        self.style = style
        if isinstance(self.options, dict):
            self.options = TransformOptions(**{
                k: v for k, v in self.options.items() if k in TransformOptions.__dataclass_fields__
            })
        if not self.doodle_png and not self.doodle_path and not self.strokes:
            raise ValueError("provide doodle_png, doodle_path, or strokes")


@dataclass
class TransformResult:
    """Outbound transform contract. Prompts never returned to clients."""

    status: str  # "ok" | "error" | "dry_run"
    style: str
    transform_version: str
    provider: str | None = None
    model: str | None = None
    image_path: str | None = None
    image_url: str | None = None
    image_base64: str | None = None
    error: str | None = None
    metadata: dict[str, Any] = field(default_factory=dict)

    def to_dict(self, *, include_image_base64: bool = True) -> dict[str, Any]:
        d = asdict(self)
        if not include_image_base64:
            d.pop("image_base64", None)
        return d

File: mobile/src/transform/contracts.ts
/**
 * Mobile-facing transform contracts — mirror of product/transform/contracts.py.
 * UI knows only TransformRequest / TransformResult. No prompts, keys, or providers.
 */

import type { StyleId } from '../models/types';

export interface TransformOptions {
  size?: number;
  dry_run?: boolean;
}

export interface StrokeJsonPoint {
  /** Schema uses [x, y] tuples. */
  0?: number;
  1?: number;
}

export interface StrokeJsonStroke {
  id: string;
  points: [number, number][];
  closed?: boolean;
  color?: string | null;
}

export interface StrokeJson {
  canvas: { width: number; height: number };
  strokes: StrokeJsonStroke[];
  meta?: Record<string, unknown>;
}

export interface TransformRequest {
  style: StyleId;
  /** PNG as base64 (no data: prefix required). */
  doodle_base64?: string;
  strokes?: StrokeJson;
  client_doodle_id?: string;
  options?: TransformOptions;
}

export type TransformStatus = 'ok' | 'error' | 'dry_run';

export interface TransformResult {
  status: TransformStatus;
  style: StyleId | string;
  transform_version: string;
  provider?: string | null;
  model?: string | null;
  image_path?: string | null;
  image_url?: string | null;
  image_base64?: string | null;
  error?: string | null;
  metadata?: Record<string, unknown>;
}

export const TRANSFORM_VERSION = 'product.mvp.v1';
